src/state/selectors/barSelectors.ts

```diff
@@ -0,0 +1,495 @@
+        acc[bar.stackId] = [];
+      }
+      acc[bar.stackId].push(bar);
+      return acc;
+    }, initialValue);
+
+    const stackedSizeList = Object.entries(groupByStack).map(([stackId, bars]) => {
+      const dataKeys = bars.map(b => b.dataKey);
+      const barSize = getBarSize(globalSize, totalSize, bars[0].barSize);
+      return { stackId, dataKeys, barSize };
+    });
+
+    const unstackedSizeList = unstackedBars.map(b => {
+      const dataKeys = [b.dataKey];
+      const barSize = getBarSize(globalSize, totalSize, b.barSize);
+      return { stackId: undefined, dataKeys, barSize };
+    });
+
+    return [...stackedSizeList, ...unstackedSizeList];
+  },
+);
+
+export const selectBarBandSize: (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+  barSettings: BarSettings,
+) => number | undefined = (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+  barSettings: BarSettings,
+): number | undefined => {
+  const layout = selectChartLayout(state);
+  const globalMaxBarSize = selectRootMaxBarSize(state);
+  const { maxBarSize: childMaxBarSize } = barSettings;
+  const maxBarSize: number = isNil(childMaxBarSize) ? globalMaxBarSize : childMaxBarSize;
+  let axis: BaseAxisWithScale, ticks: ReadonlyArray<TickItem>;
+  if (layout === 'horizontal') {
+    axis = selectAxisWithScale(state, 'xAxis', xAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'xAxis', xAxisId, isPanorama);
+  } else {
+    axis = selectAxisWithScale(state, 'yAxis', yAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'yAxis', yAxisId, isPanorama);
+  }
+  return getBandSizeOfAxis(axis, ticks, true) ?? maxBarSize ?? 0;
+};
+
+const selectAxisBandSize = (
+  state: RechartsRootState,
+  xAxisId: AxisId,
+  yAxisId: AxisId,
+  isPanorama: boolean,
+): number | undefined => {
+  const layout = selectChartLayout(state);
+  let axis: BaseAxisWithScale, ticks: ReadonlyArray<TickItem>;
+  if (layout === 'horizontal') {
+    axis = selectAxisWithScale(state, 'xAxis', xAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'xAxis', xAxisId, isPanorama);
+  } else {
+    axis = selectAxisWithScale(state, 'yAxis', yAxisId, isPanorama);
+    ticks = selectTicksOfGraphicalItem(state, 'yAxis', yAxisId, isPanorama);
+  }
+  return getBandSizeOfAxis(axis, ticks);
+};
+
+function getBarPositions(
+  barGap: string | number,
+  barCategoryGap: string | number,
+  bandSize: number,
+  sizeList: SizeList,
+  maxBarSize: number,
+): ReadonlyArray<BarWithPosition> {
+  const len = sizeList.length;
+  if (len < 1) return null;
+
+  let realBarGap = getPercentValue(barGap, bandSize, 0, true);
+
+  let result: ReadonlyArray<BarWithPosition>;
+  const initialValue: ReadonlyArray<BarWithPosition> = [];
+
+  // whether is barSize set by user
+  // Okay but why does it check only for the first element? What if the first element is set but others are not?
```

Comment:
Hmm, maybe this only accepts props from the higher `BarChart` so all items after should be the same as the first? idk

Reply:
If it's only from the root BarChart then why is this a property on every item in the sizeList? I left this comment as an opportunity for improvement.

Reply:
Does it accept the same property from both BarChart and Bar and then add the former to every item if none specified on the latter?

Just making things up but that's possible